test/testutilsio.cpp

```diff
@@ -0,0 +1,115 @@
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include <QtGlobal>
+#include <QTest>
+
+#include "base/global.h"
+#include "base/path.h"
+#include "base/utils/io.h"
+
+class TestUtilsIO final : public QObject
+{
+    Q_OBJECT
+    Q_DISABLE_COPY_MOVE(TestUtilsIO)
+
+public:
+    TestUtilsIO() = default;
+
+private slots:
+    void testReadFile() const
+    {
+        const Path testFolder = Path(QString::fromUtf8(__FILE__)).parentPath() / Path(u"testdata"_qs);
+
+        const Path size10File = testFolder / Path(u"size10.txt"_qs);
+        const QByteArray size10Data = QByteArrayLiteral("123456789\n");
+
+        {
+            const auto readResult = Utils::IO::readFile(size10File, 0);
+            QCOMPARE(readResult.has_value(), false);
+            QCOMPARE(readResult.error().status, Utils::IO::ReadError::ExceedSize);
+            QCOMPARE(readResult.error().message.isEmpty(), false);
+        }
+        {
+            const auto readResult = Utils::IO::readFile(size10File, 9);
+            QCOMPARE(readResult.has_value(), false);
+            QCOMPARE(readResult.error().status, Utils::IO::ReadError::ExceedSize);
+            QCOMPARE(readResult.error().message.isEmpty(), false);
+        }
+        {
+            const auto readResult = Utils::IO::readFile(size10File, 10);
+            QCOMPARE(readResult.has_value(), true);
+            QCOMPARE(readResult.value(), size10Data);
+        }
+        {
+            const auto readResult = Utils::IO::readFile(size10File, 11);
+            QCOMPARE(readResult.has_value(), true);
+            QCOMPARE(readResult.value(), size10Data);
+        }
+        {
+            const auto readResult = Utils::IO::readFile(size10File, -1);
+            QCOMPARE(readResult.has_value(), true);
+            QCOMPARE(readResult.value(), size10Data);
+        }
+
+        {
+            const Path nonExistFile = testFolder / Path(u".non_existent_file_1234"_qs);
+            const auto readResult = Utils::IO::readFile(nonExistFile, 1);
+            QCOMPARE(readResult.has_value(), false);
+            QCOMPARE(readResult.error().status, Utils::IO::ReadError::NotExist);
+            QCOMPARE(readResult.error().message.isEmpty(), false);
+        }
+
+#ifdef Q_OS_UNIX
+        {
+            const auto readResult = Utils::IO::readFile(Path(u"/dev/null"_qs), 10);
+            QCOMPARE(readResult.has_value(), true);
+            QCOMPARE(readResult.value().length(), 0);
+        }
+        {
+            const auto readResult = Utils::IO::readFile(Path(u"/dev/null"_qs), -1);
+            QCOMPARE(readResult.has_value(), true);
+            QCOMPARE(readResult.value().length(), 0);
+        }
+
+        {
+            const auto readResult = Utils::IO::readFile(Path(u"/dev/zero"_qs), 10);
+            QCOMPARE(readResult.has_value(), true);
+            QCOMPARE(readResult.value().length(), 0);
```

Comment:
Shouldn't it read 10 zeros?

Reply:
...or is `Utils::IO::readFile` not purposed to read files of this kind at all?

Reply:
>...or is Utils::IO::readFile not purposed to read files of this kind at all?

I intend this.